drivers/soundwire/bus.c

```diff
@@ -16,27 +16,47 @@
  * Initializes the bus instance, read properties and create child
  * devices.
  */
-int sdw_add_bus_master(struct sdw_bus *bus)
+struct sdw_master_device *
+sdw_add_bus_master(struct sdw_bus *bus, struct device *parent,
+		struct fwnode_handle *fwnode, struct sdw_link_ops *link_ops,
+		void *pdata)
 {
 	struct sdw_master_prop *prop = NULL;
+	struct sdw_master_device *md;
 	int ret;
 
+	md = sdw_master_device_add(parent,
+                                   fwnode,
+                                   link_ops,
+                                   bus->link_id,
+                                   pdata);
```

Comment:
I would have done the exact opposite. The main API is sdw_master_device_add, which calls the links that does all the inits and then you add the sdw_add_bus_master as the last stage of sdw_master_device_add().

In other words sdw_add_bus_master() is a no longer an externally visible API but an internal one.

Reply:
> I would have done the exact opposite. The main API is sdw_master_device_add, which calls the link .add callback that does all the inits and then you add the sdw_add_bus_master as the last stage of sdw_master_device_add().
> 
> In other words sdw_add_bus_master() is a no longer an externally visible API but an internal one.

I call sdw_add_bus_master() in sdw_intel_probe_controller() and sdw_add_bus_master() has almost the same parameters as sdw_master_device_add(). So as I can imagine that your proposal will be very similar to this PR, right?